spec/controllers/omniauth_callbacks_controller_spec.rb

```diff
@@ -1,62 +1,144 @@
-#         get :openid_connect
-#       end
-
-#       it 'redirects to the registration page with a flash message' do
-#         expect(flash[:notice]).to eq('Something went wrong, Please try signing-up here.')
-#         expect(response).to redirect_to(new_user_registration_path)
-#       end
-#     end
-
-#     context 'with correct credentials' do
-#       before do
-#         create(:org, managed: false, is_other: true)
-#         @org = create(:org, managed: true)
-#         @identifier_scheme = create(:identifier_scheme,
-#                                     name: 'openid_connect',
-#                                     description: 'CILogon',
-#                                     active: true,
-#                                     identifier_prefix: 'https://www.cilogon.org/')
-
-#         Rails.application.env_config['devise.mapping'] = Devise.mappings[:user]
-#         Rails.application.env_config['omniauth.auth'] = OmniAuth.config.mock_auth[:openid_connect]
-#         allow(User).to receive(:from_omniauth).and_return(user)
-#         # get :openid_connect
-#       end
-
-#       it 'links account from external credentials' do
-#         expect(flash[:notice]).to eq('Linked successfully')
-#         expect(response).to redirect_to(root_path)
-#       end
-#     end
-#   end
-# end
+
+RSpec.describe Users::OmniauthCallbacksController, type: :controller do
+  before do
+    # Enable test mode for OmniAuth
+    OmniAuth.config.test_mode = true
+
+    # Setup Devise mapping
+    @request.env["devise.mapping"] = Devise.mappings[:user]
+    create(:org, managed: false, is_other: true)
+    @org = create(:org, managed: true)
+    @identifier_scheme = create(:identifier_scheme,
+                                name: 'openid_connect',
+                                description: 'CILogon',
+                                active: true,
+                                identifier_prefix: 'https://www.cilogon.org/')
+
+    # Mock OmniAuth data for OpenID Connect with necessary info
+    OmniAuth.config.mock_auth[:openid_connect] = OmniAuth::AuthHash.new({
+      provider: 'openid_connect',
+      uid: '12345',
+      info: {
+        email: 'test@example.com',
+        first_name: 'Test',
+        last_name: 'User',
+        name: 'Test User'
+      }
+    })
+
+    # Assign the mocked authentication hash to the request environment
+    @request.env["omniauth.auth"] = OmniAuth.config.mock_auth[:openid_connect]
+  end
+
+  describe 'GET #openid_connect' do
+    let(:auth) { request.env['omniauth.auth'] }
+    let!(:identifier_scheme) { IdentifierScheme.create(name: auth.provider) }
+
+    context 'when the email is missing and the user does not exist' do
+      before do
+        # Simulate missing email
+        OmniAuth.config.mock_auth[:openid_connect].info.email = nil
+        @request.env["omniauth.auth"] = OmniAuth.config.mock_auth[:openid_connect]
+
+        def User.from_omniauth(_auth)
+          nil
+        end
```

Comment:
How does monkey patching this method work here? Will it be reset after the tests are done ? Or do we need to reset it for the rest of the tests ?

Reply:
Could we change the values we are using to get the nil value as a return instead of monkey patching this method ?